srml/contract/src/account_db.rs

```diff
@@ -26,10 +27,12 @@ use runtime_primitives::traits::Zero;
 use srml_support::{StorageMap, traits::{UpdateBalanceOutcome,
 	SignedImbalance, Currency, Imbalance}, storage::child};
 
+// TODO TODO: be careful update current doc !!
+
 pub struct ChangeEntry<T: Trait> {
 	balance: Option<BalanceOf<T>>,
-	/// In the case the outer option is None, the code_hash remains untouched, while providing `Some(None)` signifies a removing of the code in question
-	code: Option<Option<CodeHash<T>>>,
+	/// If None, the code_hash remains untouched.
+	contract: Option<NewContract<T>>,
```

Comment:
Hm, but what about `rent_allowance`? A contract should be able to alter its own `rent_allowance`. But in the current construction a contract should deal with codehash as well, right?

Reply:
indeed I forget the update of rent_allowance mecanism

Reply:
Well but do we want this mecanism to be used in the overlay ? several question then:

it seems to me changing the rent allowance in overlay or defered is same as rent wil be upgrade after first call.
so it can be provided with an extrinsic (same as restore_to)

Reply:
I am not sure if I understand you. 

So, there are going to be `ext_set_rent_allowance`/`ext_rent_allowance` functions. If a contract called `ext_set_rent_allowance` but then changes were rollbacked we would have to rollback `set_allowance` as well, right?

Reply:
My point was if we only provide ext_set_rent_allowance then it doesn't matter if it is defered to the end the transaction call or not.
But if we provide both ext_set_rent_allowance and ext_rent_allowance then fine we can put in the overlay. So ok I do that

Reply:
> My point was if we only provide `ext_set_rent_allowance` then it doesn't matter if it is defered to the end the transaction call or not.

But why? Even without `ext_rent_allowance` the differences between overlayed and immediate writes will be observable.

Consider the following sequence of actions:

1. a contract calls `ext_set_rent_allowance(0)`
2. the contract fails.

Under the overlay approach this change will be rollbacked. Under the direct write approach, however, the write gets through even though the contract has failed. This gets you an observable difference between eviction of the contract on the next block.

Reply:
no I meant differed using an extrinsic call. which are rollbacked. But if we want to provide ext_get_rent_allowance then we can't do this way.

Reply:
also I made the assumption that code_hash can only be changed in the overlay when a contract is created. we'll probably leverage this assumption when we want contract to update there code.

Thus I will make it future proof as well

Reply:
no I meant during commit in DirectAccountDb I made some assert and some were wrong if we take into account that CodeHash can be changed.

Now it is updated, though I don't the necessity to had a way to remove a contract in the accountDb.

(Also all in all I feel we designed accountDb in a different way. I tend to make the traits as safe as possible and only adapted for our needs.
On the other hand you seemed to have designed a raw overlay)